### PR TITLE
:white_check_mark: Add unit tests for some model and core classes

### PR DESCRIPTION
**Pull request by janiceblundell**
Initial unit tests for some model and core objects.
[Issue#33](https://github.com/Shopify/web-foundations-private/issues/33)

### CODE REVIEW

File: packages/core/tests/error.test.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  LogOptions,
+  LogLevel,
+} from '..';
+
+const content: Loggable = ['This is what happened', 'and this'];
+
+describe('DiagnosticError', () => {
+  const title = 'An error happened';
+  const suggestion: Loggable = 'This is what you can do about it';
+  it('defines a DiagnosticError', () => {
```
Comment (ismail-syed) [inline packages/core/tests/error.test.ts:8]: I feel like this test is doing too much and can be broken down in the a number of different tests. 

* One test for `isDiagnosticError` behaviour with all it's respective assertions
* One test for verifying it's association with the `Error` class with it's respective assertions

I don't think there's much value in having the assertions for `title`, `content`, `suggestion` since the class doesn't do anything with it. 

`expect(err.message).toMatch(title);` seems ok. Lets readers know that  the `title` property gets propagated to the Error constructor.

File: packages/core/tests/error.test.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    expect(err.title).toMatch(title);
+    expect(err.content).toMatchObject(content);
+    expect(err.suggestion).toMatch(suggestion);
+    expect(err).toBeInstanceOf(Error);
+    expect(err.name).toMatch(Error.name);
+    expect(err.message).toMatch(title);
+    expect(err.stack).toBeDefined();
+  });
+});
+describe('Loggable', () => {
```
Comment (ismail-syed) [inline packages/core/tests/error.test.ts]: I don't think this is actually testing any real source? I would just leave it out. 
Log tests should live in the [cli package](https://github.com/ismail-syed/sewing-kit/blob/main/packages/cli/src/common.ts#L266).

File: packages/model/src/base/tests/dependencies.test.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
+import {PackageJson} from '../dependencies';
+
+describe('PackageJson', () => {
+  it('loads the package.json for its own package @sewing-kit/model', () => {
```
Comment (ismail-syed) [inline packages/model/src/base/tests/dependencies.test.ts]: This test feels a bit fragile to me. Eg, if `@sewing-kit/model` every removes the `glob` dependency, this test would fail. I would just leave this test out.

File: packages/model/src/tests/workspace.test.ts [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+// Test creation of different types of workspaces
+describe('Workspace', () => {
+  const name = 'test';
+  const root = '../../';
+
+  const options: Options = {
+    name,
+    root,
+  };
+  const base = new Base(options);
```
Comment (ismail-syed) [inline packages/model/src/tests/workspace.test.ts]: There's a lot jumbled up here and it's hard to follow. I think we can break these up into their own respective `describe` blocks. 

```tsx
describe('Workspace', () => {
  describe('package', () => {});
  describe('webApp', () => {});
  describe('service', () => {});
});
```

File: packages/core/tests/error.test.ts (hunk 2 above)
Comment (janiceblundell) [reply packages/core/tests/error.test.ts]: The Log and Loggable interfaces are in core, so I put tests for them here.  This test isn't actually logging anything, the implementation of the Log function is mocked out.  I was basically just exercising the interface, and how a Log method could be defined and used.  Would it be okay to leave this here and the more e2e tests for the cli would live in the cli package and actually test logging?

File: packages/model/src/tests/workspace.test.ts [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+  it('creates a workspace for a WebApp', () => {
+    const opts: WorkspaceOptions = {
+      webApps,
+      services: emptyServices,
+      packages: emptyPackages,
+      name,
+      root,
+    };
+    const webAppWorkspace = new Workspace(opts);
+    expect(webAppWorkspace.projects[0].kind).toBe(ProjectKind.WebApp);
```
Comment (ismail-syed) [inline packages/model/src/tests/workspace.test.ts]: Nice one. The `kind` property is what we'll probably want to use as differentiation between all these different `Base` extended classes. New readers can now self discover this with these tests. 👍

File: packages/model/src/tests/workspace.test.ts [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    const opts: WorkspaceOptions = {
+      webApps: emptyWebApps,
+      services: emptyServices,
+      packages,
+      name,
+      root,
+    };
+    const packageWorkspace = new Workspace(opts);
+    expect(packageWorkspace.projects[0].kind).toBe(ProjectKind.Package);
+    expect(packageWorkspace.private).toBe(false);
```
Comment (ismail-syed) [inline packages/model/src/tests/workspace.test.ts]: Not entirely sure what `private` is intended for given how vague it's named. Perhaps @lemonmade can shine some light.

File: packages/model/src/tests/workspace.test.ts (hunk 4 above)
Comment (janiceblundell) [reply packages/model/src/tests/workspace.test.ts]: I have since created some utilities which cleans up these tests, I will update and call for review again after I have addressed the other comments as well.

File: packages/model/src/tests/workspace.test.ts (hunk 6 above)
Comment (janiceblundell) [reply packages/model/src/tests/workspace.test.ts]: What I inferred from the code is that a workspace with only a package or packages would be public, i.e packages that could be published, every other kind of workspace would be private.  Not sure if that was what was intended?

File: packages/model/src/base/tests/dependencies.test.ts (hunk 3 above)
Comment (janiceblundell) [reply packages/model/src/base/tests/dependencies.test.ts]: Yes, I should create a test package.json specifically for the test instead of using a real one.  I wanted to test the two ways of creating a PackageJson object.

File: packages/core/tests/error.test.ts (hunk 2 above)
Comment (janiceblundell) [reply packages/core/tests/error.test.ts]: I updated the test name and the test so it makes more sense.

File: packages/model/README.md [hunk 7]
```diff
@@ -0,0 +1,6 @@
+# Model
```
Comment (ismail-syed) [inline packages/model/README.md]: I think we should follow the style most of our readmes follow. 

* Package name as the header
* A brief explanation of the package. What you have is great!
* Installation instructions
* API reference and Usage examples

The READMEs in our quilt packages are pretty well documented and follow that structure fairly well. (eg. [`@shopify/with-env`](https://github.com/Shopify/quilt/tree/master/packages/with-env), [`@shopify/async`](https://github.com/Shopify/quilt/tree/master/packages/async)). There are also a few sewing-kit-next plugins with documented READMEs ([`@sewing-kit/plugin-javascript`](https://github.com/ismail-syed/sewing-kit/tree/main/packages/plugin-javascript), [`@sewing-kit/plugin-eslint`](https://github.com/ismail-syed/sewing-kit/tree/main/packages/plugin-eslint)).

File: packages/core/tests/error.test.ts [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+  Log,
+  LogOptions,
+  LogLevel,
+} from '..';
+
+const title = 'An error happened';
+const content: Loggable = ['This is what happened', 'and this'];
+const suggestion: Loggable = 'This is what you can do about it';
+
+describe('DiagnosticError', () => {
```
Comment (ismail-syed) [inline packages/core/tests/error.test.ts:7]: We generally like to have empty lines between `it` block. 

```tsx
describe('TestTheThing', () => {
  it('foo', () => {
  });

  it('bar', () => {
  });

  it('baz', () => {
  });
});
```

File: packages/core/tests/error.test.ts [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+    expect(isDiagnosticError(diagErr)).toBe(true);
+  });
+
+  it('defines an Error', () => {
+    const err = new DiagnosticError({title, content, suggestion});
+    expect(err).toBeInstanceOf(Error);
+    expect(err.message).toMatch(title);
+  });
+
+  it('can be logged', () => {
```
Comment (ismail-syed) [inline packages/core/tests/error.test.ts]: I'm still struggling to understand if this is testing anything in `DiagnosticError`. What its testing is its own self-created `log` function. I think we should just remove this test. If we're testing how the interface can be implemented, this type of test should live in the thing that's implementing the interface. In our case, I believe its the [cli package](https://github.com/ismail-syed/sewing-kit/blob/main/packages/cli/src/common.ts#L266).

File: packages/core/tests/target.test.ts [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+  Runtime,
+  WorkspaceOptions,
+  Workspace,
+} from '../../model';
+import {
+  createTestWebAppWorkspace,
+  createTestServiceWorkspace,
+  createTestPackageWorkspace,
+  createTestComboWorkspace,
+} from '../../model/tests/utilities';
```
Comment (ismail-syed) [inline packages/core/tests/target.test.ts:17]: Reaching in the other packages test utilities is generally discouraged. Ideally, this should live in its own testing packages like `@sewing-kit/testing` but for now, maybe we can move these into the root `tests/utilities`

File: packages/model/tests/dependencies.test.ts [hunk 11]
```diff
@@ -0,0 +1,33 @@
+import {resolve} from 'path';
+import {PackageJson} from '../src/base/dependencies';
+
+const fakeProject = resolve(__dirname, 'project');
+
+jest.mock(
```
Comment (ismail-syed) [inline packages/model/tests/dependencies.test.ts:6]: yay for mocking!

File: packages/model/tests/workspace.test.ts [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+} from './utilities';
+
+// Test creation of different types of workspaces
+describe('Workspace', () => {
+  const name = 'test';
+  const root = '../../';
+
+  it('creates a workspace for a WebApp', () => {
+    const webAppWorkspace = createTestWebAppWorkspace(name, root);
+    expect(webAppWorkspace.name).toBe(name);
```
Comment (ismail-syed) [inline packages/model/tests/workspace.test.ts]: I don't think there's much value in the `expect(thing.name).toBe(name);` in here or the other following tests. Lets remove it. Setting `name` is the responsibility of the `Base` class. That's not what we are testing here so it's just redundant assertions.

**Review by ismail-syed — APPROVED** [review]
Just left a few small comments. Aside from those, this is good to go. I believe Chris is on vacation so we won't be able to get this merged until he's back. Even more reason for me to pull this codebase into the org 😅

**Comment by ismail-syed** [comment]
Closing in favor of https://github.com/Shopify/sewing-kit-next/pull/45
